File: trunk/main.net/src/Coherence.Tools/Config/Configuration.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Configuration;

using Common.Logging;

namespace Seovic.Config
{
    /// <summary>
    /// Provides centralized access to Coherence Tools configuration info.
    /// </summary>
    /// <remarks>
    /// You can modify configuration settings by editing 
    /// <c>Coherence.Tools.dll.config</c> configuration file, which should
    /// be colocated in the same directory with <c>Coherence.Tools.dll</c>. 
    /// </remarks>
    /// <author>Aleksandar Seovic  2010.02.05</author>
    public class Configuration
    {
        #region Constructors

        /// <summary>
        /// Singleton constructor.
        /// </summary>
        private Configuration()
        {
            m_config = LoadConfiguration();
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Get a default expression type.
        /// </summary>
        /// <returns>Default expression type.</returns>
        public static Type GetDefaultExpressionType()
        {
            return Type.GetType(Instance.m_config[EXPRESSION_TYPE]);
        }

        /// <summary>
        /// Get a default extractor type.
        /// </summary>
        /// <returns>Default extractor type.</returns>
        public static Type GetDefaultExtractorType()
        {
            return Type.GetType(Instance.m_config[EXTRACTOR_TYPE]);
        }

        /// <summary>
        /// Get a default updater type.
        /// </summary>
        /// <returns>Default updater type.</returns>
        public static Type GetDefaultUpdaterType()
        {
            return Type.GetType(Instance.m_config[UPDATER_TYPE]);
        }

        /// <summary>
        /// Get a default condition type.
        /// </summary>
        /// <returns>Default condition type.</returns>
        public static Type GetDefaultConditionType()
        {
            return Type.GetType(Instance.m_config[CONDITION_TYPE]);
        }

        /// <summary>
        /// Return the name of the default script language.
        /// </summary>
        /// <returns>
        /// The name of the default script language.
        /// </returns>
        public static string GetDefaultScriptLanguage()
        {
            return Instance.m_config[SCRIPT_LANGUAGE];
        }

        /// <summary>
        /// Get sequence generator type.
        /// </summary>
        /// <returns>Sequence generator type.</returns>
        public static Type GetSequenceGeneratorType()
        {
            return Type.GetType(Instance.m_config[SEQUENCE_GENERATOR_TYPE]);
        }

        /// <summary>
        /// Return the name of the sequence cache.
        /// </summary>
        /// <returns>
        /// The name of the sequence cache.
        /// </returns>
        public static string GetSequenceCacheName()
        {
            return Instance.m_config[SEQUENCE_CACHE_NAME];
        }

        #endregion

        #region Helper methods

        /// <summary>
        /// Loads default values from a application configuration file.
        /// </summary>
        /// <returns>
        /// Dictinary containing configuration values.
        /// </returns>
        private static IDictionary<string, string> LoadConfiguration()
        {
            IDictionary<string, string> props = new Dictionary<string, string>();
            props.Add(EXPRESSION_TYPE, DEFAULT_EXPRESSION_TYPE);
            props.Add(EXTRACTOR_TYPE,  DEFAULT_EXTRACTOR_TYPE);
            props.Add(UPDATER_TYPE,    DEFAULT_UPDATER_TYPE);
            props.Add(CONDITION_TYPE,  DEFAULT_CONDITION_TYPE);
            props.Add(SCRIPT_LANGUAGE, DEFAULT_SCRIPT_LANGUAGE);

            props.Add(SEQUENCE_GENERATOR_TYPE, DEFAULT_SEQUENCE_GENERATOR_TYPE);
            props.Add(SEQUENCE_CACHE_NAME,     DEFAULT_SEQUENCE_CACHE_NAME);

            try
            {
                NameValueCollection config = ConfigurationManager.AppSettings;
                foreach (string key in config.Keys)
                {
                    props[key] = config[key];
                }
            }
            catch (Exception)
            {
                // should never happen
                log.Warn("Configuration file Coherence.Tools.dll.config"
                         + " is missing. Using hardcoded defaults: \n" + props);
            }
            return props;
        }

        #endregion

        #region Constants
        
        #region Configuration property keys

        private const string EXPRESSION_TYPE = "expression.type";
        private const string EXTRACTOR_TYPE  = "extractor.type";
        private const string UPDATER_TYPE    = "updater.type";
        private const string CONDITION_TYPE  = "condition.type";
        private const string SCRIPT_LANGUAGE = "script.language";

        private const string SEQUENCE_GENERATOR_TYPE = "sequence.generator.type";
        private const string SEQUENCE_CACHE_NAME     = "sequence.cache.name";

        #endregion

        #region Configuration default values

        private const string DEFAULT_EXPRESSION_TYPE = "Seovic.Core.Expression.SpelExpression, Coherence.Tools";
        private const string DEFAULT_EXTRACTOR_TYPE  = "Seovic.Core.Extractor.ExpressionExtractor, Coherence.Tools";
        private const string DEFAULT_UPDATER_TYPE    = "Seovic.Core.Updater.ExpressionUpdater, Coherence.Tools";
        private const string DEFAULT_CONDITION_TYPE  = "Seovic.Core.Condition.ExpressionCondition, Coherence.Tools";
        private const string DEFAULT_SCRIPT_LANGUAGE = "javascript";

        private const string DEFAULT_SEQUENCE_GENERATOR_TYPE = "Seovic.Core.Coherence.Identity.CoherenceSequenceGenerator, Coherence.Tools";
        private const string DEFAULT_SEQUENCE_CACHE_NAME     = "sequences";

        #endregion

        #endregion

        #region Data members

        /// <summary>
        /// Singleton instance.
        /// </summary>
        private static readonly Configuration Instance = new Configuration();

        /// <summary>
        /// Logger for this class.
        /// </summary>
        private static readonly ILog log = LogManager.GetLogger(typeof(Configuration));

        /// <summary>
        /// Configuration settings.
        /// </summary>
        private readonly IDictionary<string, string> m_config;

        #endregion
    }
}

File: main/src/com/seovic/config/Configuration.java
/*
 * Copyright 2009 Aleksandar Seovic
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.seovic.config;


import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import java.util.Map;
import java.util.HashMap;
import java.util.Properties;


/**
 * Provides centralized access to Coherence Tools configuration info.
 * <p/>
 * You can modify configuration settings by editing
 * <tt>coherence-tools.properties</tt> configuration file, which should
 * be located in the classpath root.
 * 
 * @author Aleksandar Seovic  2010.02.05
 */
public class Configuration
    {
    // ---- constructors ----------------------------------------------------

    /// <summary>
    /// Singleton constructor.
    /// </summary>
    private Configuration()
    {
        m_config = loadConfiguration();
    }


    // ---- public methods --------------------------------------------------

    /**
     * Get a default expression type.
     *
     * @return default expression type
     *
     * @throws ClassNotFoundException  if specified class cannot be found
     */
    public static Class getDefaultExpressionType()
            throws ClassNotFoundException
        {
        return Class.forName(s_instance.m_config.get(EXPRESSION_TYPE));
        }

    /**
     * Get a default extractor type.
     *
     * @return default extractor type
     *
     * @throws ClassNotFoundException  if specified class cannot be found
     */
    public static Class getDefaultExtractorType()
            throws ClassNotFoundException
        {
        return Class.forName(s_instance.m_config.get(EXTRACTOR_TYPE));
        }

    /**
     * Get a default updater type.
     *
     * @return default updater type
     *
     * @throws ClassNotFoundException  if specified class cannot be found
     */
    public static Class getDefaultUpdaterType()
            throws ClassNotFoundException
        {
        return Class.forName(s_instance.m_config.get(UPDATER_TYPE));
        }

    /**
     * Get a default condition type.
     *
     * @return default condition type
     *
     * @throws ClassNotFoundException  if specified class cannot be found
     */
    public static Class getDefaultConditionType()
            throws ClassNotFoundException
        {
        return Class.forName(s_instance.m_config.get(CONDITION_TYPE));
        }

    /**
     * Get the name of the default script language.
     *
     * @return the name of the default script language
     */
    public static String getDefaultScriptLanguage()
        {
        return s_instance.m_config.get(SCRIPT_LANGUAGE);
        }

    /**
     * Get sequence generator type.
     *
     * @return sequence generator type
     *
     * @throws ClassNotFoundException  if specified class cannot be found
     */
    public static Class getSequenceGeneratorType()
            throws ClassNotFoundException
        {
        return Class.forName(s_instance.m_config.get(SEQUENCE_GENERATOR_TYPE));
        }

    /**
     * Get the name of the sequence cache.
     *
     * @return the name of the sequence cache
     */
    public static String getSequenceCacheName()
        {
        return s_instance.m_config.get(SEQUENCE_CACHE_NAME);
        }

    /**
     * Get the name of the job scheduler cache.
     *
     * @return the name of the job scheduler cache
     */
    public static String getSchedulerCacheName()
        {
        return s_instance.m_config.get(SCHEDULER_CACHE_NAME);
        }

    // ---- helper methods --------------------------------------------------

    /// <summary>
    /// Loads default values from a application configuration file.
    /// </summary>
    /// <returns>
    /// Dictionary containing configuration values.
    /// </returns>
    private static Map<String, String> loadConfiguration()
        {
        Map<String, String> props = new HashMap<String, String>();
        props.put(EXPRESSION_TYPE, DEFAULT_EXPRESSION_TYPE);
        props.put(EXTRACTOR_TYPE,  DEFAULT_EXTRACTOR_TYPE);
        props.put(UPDATER_TYPE,    DEFAULT_UPDATER_TYPE);
        props.put(CONDITION_TYPE,  DEFAULT_CONDITION_TYPE);
        props.put(SCRIPT_LANGUAGE, DEFAULT_SCRIPT_LANGUAGE);

        props.put(SEQUENCE_GENERATOR_TYPE, DEFAULT_SEQUENCE_GENERATOR_TYPE);
        props.put(SEQUENCE_CACHE_NAME,     DEFAULT_SEQUENCE_CACHE_NAME);

        try
            {
            Properties config = new Properties();
            config.load(Configuration.class.getClassLoader()
                    .getResourceAsStream("coherence-tools.properties"));
            for (String propertyName : config.stringPropertyNames())
                {
                props.put(propertyName, config.getProperty(propertyName));
                }
            }
        catch (Exception e)
            {
            // should never happen, as default file is embedded within JAR
            s_log.warn("Configuration file coherence-tools.properties"
                     + " is missing. Using hardcoded defaults: \n" + props);
            }
        return props;
    }


    // ---- constants -------------------------------------------------------

    // configuration property keys
    private static final String EXPRESSION_TYPE = "expression.type";
    private static final String EXTRACTOR_TYPE  = "extractor.type";
    private static final String UPDATER_TYPE    = "updater.type";
    private static final String CONDITION_TYPE  = "condition.type";
    private static final String SCRIPT_LANGUAGE = "script.language";

    private static final String SEQUENCE_GENERATOR_TYPE = "sequence.generator.type";
    private static final String SEQUENCE_CACHE_NAME     = "sequence.cache.name";

    private static final String SCHEDULER_CACHE_NAME     = "scheduler.cache.name";

    // default values
    private static final String DEFAULT_EXPRESSION_TYPE = "com.seovic.core.expression.SpelExpression";
    private static final String DEFAULT_EXTRACTOR_TYPE  = "com.seovic.core.extractor.ExpressionExtractor";
    private static final String DEFAULT_UPDATER_TYPE    = "com.seovic.core.updater.ExpressionUpdater";
    private static final String DEFAULT_CONDITION_TYPE  = "com.seovic.core.condition.ExpressionCondition";
    private static final String DEFAULT_SCRIPT_LANGUAGE = "javascript";

    private static final String DEFAULT_SEQUENCE_GENERATOR_TYPE = "com.seovic.identity.SimpleSequenceGenerator";
    private static final String DEFAULT_SEQUENCE_CACHE_NAME     = "sequences";


    // ---- data members ----------------------------------------------------

    /**
     * Singleton instance.
     */
    private static final Configuration s_instance = new Configuration();
    
    /**
     * Logger for this class.
     */
    private static final Log s_log = LogFactory.getLog(Configuration.class);

    /**
     * Configuration settings.
     */
    private final Map<String, String> m_config;
    }
